### PR TITLE
Add initial support for classes getter and setters

### PR DESCRIPTION
**Pull request by gabryon99**
This pull request is in WIP. Currently, it implements the following features:

- [x] Getters for 'virtual' fields, such as the example shown below.
- [x] Fields setters.

Simple test-case for custom getter:
```kotlin
class IntWrapper(val n: Int) {
    val succ: Int get() {
        return n + 1
    }
}

fun testGetters() {
    val wrapper = IntWrapper(42)
    val succ = wrapper.succ
}
```

Simple test-case for predefined setter:
```kotlin
class Person(var age: Int)

fun testSetter() {
    val person = Person(19)
    person.age = person.age + 1
}
```

### CODE REVIEW

File: plugins/formal-verification/testData/diagnostics/no_contracts/classes_getters.kt
Comment (22slin22) [inline plugins/formal-verification/testData/diagnostics/no_contracts/classes_getters.kt:1]: Maybe you can move the test of accessing a field with a backing field from `classes.kt` to here.

File: plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/conversion/StmtConversionVisitor.kt [hunk 1]
```diff
@@ -116,24 +122,38 @@ class StmtConversionVisitor : FirVisitor<Exp, StmtConversionContext>() {
...
                     // Access the field and assign the value to a newly created anonymous variable.
-                    data.addStatement(Stmt.assign(anon.toLocalVar(), fieldAccess))
-                    // Exhale permission for the field and return the anonymous variable value
-                    data.addStatement(Stmt.Exhale(accPred))
-                    return anon.toLocalVar()
+                    inhaleAndExhaleStatement(
+                        AccessPredicate.FieldAccessPredicate(fieldAccess, PermExp.FullPerm()),
+                        Stmt.assign(anon.toLocalVar(), fieldAccess),
+                        data
+                    )
```
Comment (22slin22) [inline plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/conversion/StmtConversionVisitor.kt:153]: This can be handled nicely with the new `StmtWithResultConversionContext`. Have a look at this function in the current `formal-verification` branch.

File: plugins/formal-verification/testData/diagnostics/no_contracts/classes_getters.kt
Comment (22slin22) [inline plugins/formal-verification/testData/diagnostics/no_contracts/classes_getters.kt:1]: Did you try if your approach also works for more complex settings likes `foo.bar.baz`?

**Review by 22slin22 — CHANGES_REQUESTED** [review]

**Comment by 22slin22** [comment]
In general you could also split the PR up into multiple ones (one adding custom getters, one setting field values and another for custom setters). That way you will also not run into that many merge conflicts.

File: plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/conversion/StmtConversionVisitor.kt [hunk 2]
```diff
@@ -106,6 +111,7 @@ class StmtConversionVisitor : FirVisitor<Exp, StmtConversionContext>() {
         return cvar?.toLocalVar() ?: UnitDomain.element
     }
 
+    @OptIn(SymbolInternals::class)
```
[the remark this answers is not in the document]
Comment (gabryon99) [reply plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/conversion/StmtConversionVisitor.kt:114]: This is annotation is used to access the `fir` field in the line (140):
```kotlin
val getter: FirPropertyAccessor = symbol.getterSymbol!!.fir
```

**Comment by gabryon99** [comment]
> In general you could also split the PR up into multiple ones (one adding custom getters, one setting field values and another for custom setters). That way you will also not run into that many merge conflicts.

Thanks for the feedback! :)

File: plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/conversion/StmtConversionVisitor.kt [hunk 3]
```diff
@@ -281,18 +302,51 @@ object StmtConversionVisitor : FirVisitor<Exp, StmtConversionContext<ResultTrack
...
-        val rValueType = data.embedType(variableAssignment.rValue)
-        data.addStatement(Stmt.assign(convertedLValue, convertedRValue.convertType(rValueType, lValueType)))
+
+        val isClassField = if (variableAssignment.lValue is FirPropertyAccessExpression) {
+            (variableAssignment.lValue as FirPropertyAccessExpression).dispatchReceiver !is FirNoReceiverExpression
+        } else {
+            false
+        }
+
+        if (isClassField) {
```
Comment (jesyspa) [inline plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/conversion/StmtConversionVisitor.kt:321]: Can the common code here be split out into a helper function?  (It would be nice if that code could *also* be shared with the `PropertyAccess` visitor, since there's clearly a lot of overlap.)

File: plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/embeddings/MethodSignatureEmbedding.kt [hunk 4]
```diff
@@ -21,7 +22,10 @@ class MethodSignatureEmbedding(
     val params: List<VariableEmbedding>,
     val returnType: TypeEmbedding,
 ) {
-    val name = symbol.callableId.embedName()
+    val name = when (symbol) {
+        is FirPropertyAccessorSymbol -> symbol.propertySymbol.callableId.embedName()
```
Comment (jesyspa) [inline plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/embeddings/MethodSignatureEmbedding.kt:26]: Is this correct?  Doesn't this encode both the getter and the setter using the same name?
I would be more inclined to put this logic into `embedName`.

File: plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/conversion/StmtConversionVisitor.kt [hunk 5]
```diff
@@ -281,18 +302,51 @@ object StmtConversionVisitor : FirVisitor<Exp, StmtConversionContext<ResultTrack
...
+
+            val symbol = lValue.symbol!! as FirPropertySymbol
+
+            val varEmbedding = VariableEmbedding(symbol.callableId.embedName(), lValueType)
+            val fieldAccess = Exp.FieldAccess(receiver, varEmbedding.toField())
+
+            val convertedRValue = data.convert(variableAssignment.rValue)
+            val rValueType = data.embedType(variableAssignment.rValue)
+
+            data.withoutResult {
```
Comment (jesyspa) [inline plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/conversion/StmtConversionVisitor.kt:335]: This doesn't support non-trivial properties, right?  Perhaps there should be a check for them somewhere.

File: plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/conversion/StmtConversionVisitor.kt [hunk 6]
```diff
@@ -281,18 +302,51 @@ object StmtConversionVisitor : FirVisitor<Exp, StmtConversionContext<ResultTrack
         return UnitDomain.element
     }
 
+    @OptIn(DfaInternals::class)
```
Comment (jesyspa) [inline plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/conversion/StmtConversionVisitor.kt:305]: What is the DFA?  Why do we need the internals?

File: plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/conversion/StmtConversionVisitor.kt [hunk 7]
```diff
@@ -120,8 +125,24 @@ object StmtConversionVisitor : FirVisitor<Exp, StmtConversionContext<ResultTrack
...
+                }
+
+                val receiver = data.convert(propertyAccessExpression.dispatchReceiver)
+                val anon = data.newAnonVar(varEmbedding.type)
+                data.addDeclaration(anon.toLocalVarDecl())
+
+                // Are we accessing a class field backed by a concrete field?
+                if (!symbol.hasBackingField) {
+                    val getter: FirPropertyAccessor = symbol.getterSymbol!!.fir
+                    val method = data.embedFunction(getter.symbol)
```
[the remark this answers is not in the document]
Comment (gabryon99) [reply plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/conversion/StmtConversionVisitor.kt:137]: Yes, this allows to access the corresponding associated function

File: plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/conversion/StmtConversionVisitor.kt [hunk 8]
```diff
@@ -120,8 +125,24 @@ object StmtConversionVisitor : FirVisitor<Exp, StmtConversionContext<ResultTrack
                 val varEmbedding = data.getVariableEmbedding(symbol.callableId.embedName(), type)
                 if (symbol.isLocal) {
                     return varEmbedding.toLocalVar()
+                }
+
+                val receiver = data.convert(propertyAccessExpression.dispatchReceiver)
+                val anon = data.newAnonVar(varEmbedding.type)
```
[the remark this answers is not in the document]
Comment (gabryon99) [reply plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/conversion/StmtConversionVisitor.kt:131]: You're right! It seems a leftover of the rebasing

File: plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/conversion/StmtConversionVisitor.kt (hunk 6 above)
Comment (gabryon99) [reply plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/conversion/StmtConversionVisitor.kt]: Accessing the `symbol` property was requesting an `OptIn` for internals. I've pushed a fix addressing this!

File: plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/conversion/StmtConversionVisitor.kt (hunk 7 above)
[the remark this answers is not in the document]
Comment (jesyspa) [reply plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/conversion/StmtConversionVisitor.kt:133]: In which case do they differ?  Running locally, they always appear to be the same object.

File: plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/conversion/StmtConversionVisitor.kt [hunk 9]
```diff
@@ -285,14 +302,40 @@ object StmtConversionVisitor : FirVisitor<Exp, StmtConversionContext<ResultTrack
...
+            val receiver = data.convert(lValue.dispatchReceiver)
+
+
+            val varEmbedding = VariableEmbedding(lValueSymbol.callableId.embedName(), lValueType)
+            val fieldAccess = Exp.FieldAccess(receiver, varEmbedding.toField())
+
+            val convertedRValue = data.convert(variableAssignment.rValue)
+            val rValueType = data.embedType(variableAssignment.rValue)
+
+            data.withoutResult {
```
Comment (jesyspa) [inline plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/conversion/StmtConversionVisitor.kt:324]: By the way, what does `withoutResult` accomplish here?

File: plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/conversion/StmtConversionVisitor.kt (hunk 7 above)
[the remark this answers is not in the document]
Comment (gabryon99) [reply plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/conversion/StmtConversionVisitor.kt:133]: In a previous version of the PR I needed the getter's body, now this is not the case anymore.

TLDR; they are the same object I can use the `getterSymbol`

File: plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/conversion/StmtConversionVisitor.kt (hunk 5 above)
Comment (gabryon99) [reply plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/conversion/StmtConversionVisitor.kt]: The latest commit performs this check!

File: plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/conversion/StmtConversionVisitor.kt (hunk 9 above)
Comment (gabryon99) [reply plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/conversion/StmtConversionVisitor.kt]: Just remove it because it didn't add value

File: plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/conversion/StmtConversionVisitor.kt [hunk 10]
```diff
@@ -287,14 +300,45 @@ object StmtConversionVisitor : FirVisitor<Exp, StmtConversionContext<ResultTrack
...
+            val lValueSymbol = lValue.calleeSymbol as FirPropertySymbol
+            val lValueType = data.embedType(variableAssignment.lValue)
+            val receiver = data.convert(lValue.dispatchReceiver)
+
+            val convertedRValue = data.convert(variableAssignment.rValue)
+            val rValueType = data.embedType(variableAssignment.rValue)
+
+            // According to the setter symbol, we should invoke a method, or assigning a value
+            val setterSymbol = lValueSymbol.setterSymbol!!
+            if (setterSymbol.hasBody) {
```
Comment (jesyspa) [inline plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/conversion/StmtConversionVisitor.kt:320]: Why do we decide what to do based on whether the setter has a body here, but use the existence of a backing field elsewhere?

File: plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/conversion/StmtConversionVisitor.kt [hunk 11]
```diff
@@ -355,6 +399,16 @@ object StmtConversionVisitor : FirVisitor<Exp, StmtConversionContext<ResultTrack
     private val FirResolvable.calleeNamedFunctionSymbol: FirNamedFunctionSymbol
         get() = calleeReference.toResolvedNamedFunctionSymbol()!!
 
+    @OptIn(SymbolInternals::class)
+    private val FirPropertyAccessorSymbol.hasBody: Boolean
+        get() = this.fir.body != null
+
+    private val FirVariableAssignment.isAssigningToClassField: Boolean
```
Comment (jesyspa) [inline plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/conversion/StmtConversionVisitor.kt:406]: I think this could be an extension property on the `lValue` instead of on the assignment.

File: plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/embeddings/NameEmbeddings.kt [hunk 12]
```diff
@@ -31,19 +33,51 @@ data class ClassName(val packageName: FqName, val className: Name) : MangledName
...
+        GETTER, SETTER;
+
+        override fun toString(): String = when (this) {
+            GETTER -> "getter"
+            SETTER -> "setter"
+        }
+    }
+
+    override val mangled: String
+        get() = "${super.mangled}_$accessor"
```
Comment (jesyspa) [inline plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/embeddings/NameEmbeddings.kt:52]: I think this could cause name conflicts.

File: plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/conversion/StmtConversionVisitor.kt [hunk 13]
```diff
@@ -287,14 +300,45 @@ object StmtConversionVisitor : FirVisitor<Exp, StmtConversionContext<ResultTrack
...
+                data.addStatement(Stmt.Inhale(accPred))
+                data.addStatement(Stmt.assign(fieldAccess, convertedRValue.convertType(rValueType, lValueType)))
+                data.addStatement(Stmt.Exhale(accPred))
+            }
+
+        } else {
+            val convertedLValue = data.convert(variableAssignment.lValue)
+            val lValueType = data.embedType(variableAssignment.lValue)
+            val convertedRValue = data.convert(variableAssignment.rValue)
+            val rValueType = data.embedType(variableAssignment.rValue)
```
Comment (22slin22) [inline plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/conversion/StmtConversionVisitor.kt:338]: You are computing those values in both branches so they could be moved in front of the if.

**Review by jesyspa — COMMENTED** [review]
This PR adds a number of method calls; I suspect that we want to inline these when possible.  Perhaps there's some function call helper function we should add that all these places should delegate to? @francescoo22 might have thoughts on how best to do this.

File: plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/conversion/StmtConversionVisitor.kt (hunk 10 above)
Comment (22slin22) [reply plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/conversion/StmtConversionVisitor.kt:320]: Is it even possible that this occurs? I would imagine that if there was no body then `lValueSymbol.setterSymbol` is just `null`. Or does that indicate something else?

File: plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/embeddings/NameEmbeddings.kt [hunk 14]
```diff
@@ -31,19 +33,51 @@ data class ClassName(val packageName: FqName, val className: Name) : MangledName
         get() = "pkg_${packageName.asString()}\$class_${className.asString()}"
 }
 
-data class ClassMemberName(val className: ClassName, val name: Name) : MangledName {
+open class ClassMemberName(val className: ClassName, val name: Name) : MangledName {
```
Comment (22slin22) [inline plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/embeddings/NameEmbeddings.kt:36]: I think we always want `MangledName`s to be data classes since we want equality be derived from the name fields instead of referential equality. To make sure no one instantiates this directly make it an interface or an abstract class.

File: plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/embeddings/NameEmbeddings.kt [hunk 15]
```diff
@@ -31,19 +33,51 @@ data class ClassName(val packageName: FqName, val className: Name) : MangledName
...
     override val mangled: String
         get() = "${className.mangled}\$member_${name.asString()}"
 }
 
-/** Global name not associated with a class.
+class ClassMemberAccessor(className: ClassName, name: Name, val accessor: AccessorType) : ClassMemberName(className, name) {
```
Comment (22slin22) [inline plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/embeddings/NameEmbeddings.kt:41]: Should be a data class (see above)

File: plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/embeddings/NameEmbeddings.kt (hunk 12 above)
Comment (22slin22) [reply plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/embeddings/NameEmbeddings.kt:52]: Yes

File: plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/embeddings/NameEmbeddings.kt [hunk 16]
```diff
@@ -31,19 +33,51 @@ data class ClassName(val packageName: FqName, val className: Name) : MangledName
...
+    enum class AccessorType {
```
Comment (22slin22) [inline plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/embeddings/NameEmbeddings.kt:42]: Instead of going the way and making an enum for this why not just make two data classes `ClassMemberSetter` and `ClassMemberGetter`? Would this simplify the code a bit?

File: plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/conversion/StmtConversionVisitor.kt (hunk 10 above)
Comment (gabryon99) [reply plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/conversion/StmtConversionVisitor.kt:320]: @22slin22 this is needed to distinguish the cases like:
```kotlin
class Foo(var b: Int) {
    var a: Int = 0
       set(value) { field = value + 1}
}
```

In this example `b` does not have a body, but when we set it we should use a field assignment:
```viper
inhale acc(f.b)
f.b = 1
exhale acc(f.b)
```

Instead, when the setter has a body, we have to invoke the method corresponding to it:

```viper
anon = Foo_a_setter(n)
```

File: plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/conversion/ProgramConverter.kt [hunk 17]
```diff
@@ -93,9 +94,13 @@ class ProgramConverter(val session: FirSession, override val config: PluginConfi
         val params = symbol.valueParameterSymbols.map {
             VariableEmbedding(it.embedName(), embedType(it.resolvedReturnType))
         }
-        val receiver = symbol.receiverType?.let { VariableEmbedding(ThisReceiverName, embedType(it)) }
```
Comment (jesyspa) [inline plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/conversion/ProgramConverter.kt:96]: Why not handle this through the `receiverType` extension?

File: plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/conversion/StmtConversionContext.kt [hunk 18]
```diff
@@ -26,7 +26,9 @@ interface StmtConversionContext<out RTC : ResultTrackingContext> : MethodConvers
     fun convertAndStore(exp: FirExpression): Exp.LocalVar
 
     fun newBlock(): StmtConversionContext<RTC>
+
```
Comment (jesyspa) [inline plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/conversion/StmtConversionContext.kt:29]: Why add newlines here?

File: plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/conversion/StmtConversionVisitor.kt [hunk 19]
```diff
@@ -366,6 +415,16 @@ object StmtConversionVisitor : FirVisitor<Exp, StmtConversionContext<ResultTrack
     private val FirResolvable.calleeNamedFunctionSymbol: FirNamedFunctionSymbol
         get() = calleeReference.toResolvedNamedFunctionSymbol()!!
 
+    @OptIn(SymbolInternals::class)
+    private val FirPropertyAccessorSymbol.hasBody: Boolean
+        get() = this.fir.body != null
+
+    private val FirExpression.isAssigningToClassField: Boolean
```
Comment (jesyspa) [inline plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/conversion/StmtConversionVisitor.kt:422]: Perhaps `isClassPropertyAccess` would be a better name?

File: plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/conversion/StmtConversionVisitor.kt (hunk 10 above)
Comment (jesyspa) [reply plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/conversion/StmtConversionVisitor.kt:337]: I think for this (and also the corresponding getter code), Ilya's opinion on how to get the correct behaviour would be valuable.

**Review by 22slin22 — APPROVED** [review]

File: plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/conversion/StmtConversionVisitor.kt [hunk 20]
```diff
@@ -298,14 +312,49 @@ object StmtConversionVisitor : FirVisitor<Exp, StmtConversionContext<ResultTrack
...
+            val lValue = variableAssignment.lValue as FirPropertyAccessExpression
+            val lValueSymbol = lValue.calleeSymbol as FirPropertySymbol
+            val receiver = data.convert(lValue.dispatchReceiver!!)
+            /*
+            According to the setter symbol, we should invoke a method, or assigning a value.
+
+            Setters are only relevant for variable properties.
+            Compared to the getter's, the setter symbols require more information about the backing field.
+            If a custom setter is defined on a variable property. this also exposes a backing field.
+            To check the existence of a custom setter, we can look for a function body in a `setterSymbol`.
```
Comment (jesyspa) [inline plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/conversion/StmtConversionVisitor.kt:334]: I'm struggling a bit to understand this.  When you say `variable properties`, what kind of properties don't you include there?  A single overview with what kinds of getters and setters there are and how they are encoded in the FIR would probably work better here.

It helps to write this in a declarative, not procedural style: it's not about what _we_ should do, but about the rules that govern this data structure.

File: plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/conversion/StmtConversionVisitor.kt [hunk 21]
```diff
@@ -298,14 +312,49 @@ object StmtConversionVisitor : FirVisitor<Exp, StmtConversionContext<ResultTrack
...
+            val setterSymbol = lValueSymbol.setterSymbol!!
+            if (setterSymbol.hasBody) {
+                // Since a custom setter has been defined, we should generate a statement to invoke a method call
+                val method = data.embedFunction(setterSymbol)
+                data.withResult(UnitTypeEmbedding) {
+                    data.addStatement(method.toMethodCall(listOf(receiver, convertedRValue), this.resultCtx.resultVar))
+                }
+            } else {
+                // No custom setters have been defined, we can assign the fields normally.
+                val varEmbedding = VariableEmbedding(lValueSymbol.callableId.embedName(), lValueType)
```
Comment (jesyspa) [inline plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/conversion/StmtConversionVisitor.kt:345]: I think this should be constructed via the context now.  Could you add a test where a property is used in an inline function?

**Review by jesyspa — CHANGES_REQUESTED** [review]
Blocked on Ilya's feedback on how to identify custom properties correctly, I think.

**Review by jesyspa — APPROVED** [review]
Looked over via call, looks good now
